docs/validator.go

```diff
@@ -0,0 +1,164 @@
+ * limitations under the License.
+ */
+
+package main
+
+import (
+	"fmt"
+	"io/ioutil"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"strings"
+	"text/template"
+
+	"github.com/a8m/mark"
+	"github.com/minio/cli"
+)
+
+var templateMain = `
+// +build ignore
+
+package main
+
+import (
+    "fmt"
+
+    "github.com/minio/minio-go"
+)
+
+func main() {
+        // Use a secure connection.
+        ssl := true
+
+        // Initialize minio client object.
+        minioClient, err := minio.New("play.minio.io:9000", "Q3AM3UQ867SPQQA43P2F", "zuf+tfteSlswRu7BJ86wekitnifILbZam1KYY3TG", ssl)
+        if err != nil {
+                fmt.Println(err)
+                return
+        }
+
+        {{.Text}}
+
+}`
+
+func init() {
+	// Validate go binary.
+	if _, err := exec.LookPath("go"); err != nil {
+		panic(err)
+	}
+}
+
+var globalFlags = []cli.Flag{
+	cli.StringFlag{
+		Name:  "m",
+		Value: "API.md",
+		Usage: "Path to API.md",
+	},
+	cli.IntFlag{
+		Name:  "skip",
+		Value: 2,
+		Usage: "Skip entries before validating the code.",
+	},
+}
+
+func validatorAction(ctx *cli.Context) error {
+	if !ctx.IsSet("m") {
+		return nil
+	}
+	docPath := ctx.String("m")
+	data, err := ioutil.ReadFile(docPath)
+	if err != nil {
+		return err
+	}
+
+	skipEntries := ctx.Int("skip")
+	m := mark.New(string(data), &mark.Options{
+		Gfm: true, // Github markdown support is enabled by default.
+	})
+
+	t := template.Must(template.New("examples").Parse(templateMain))
+	tmpDir, err := ioutil.TempDir("", "md-verifier")
+	if err != nil {
+		return err
+	}
+	defer os.RemoveAll(tmpDir)
+
+	fmt.Println("About to run validation tests in", tmpDir)
+	entryN := 1
+	for i := mark.NodeText; i < mark.NodeCheckbox; i++ {
+		if mark.NodeCode != mark.NodeType(i) {
+			m.AddRenderFn(mark.NodeType(i), func(node mark.Node) (s string) {
+				return ""
+			})
+			continue
+		}
+		m.AddRenderFn(mark.NodeCode, func(node mark.Node) (s string) {
+			p, ok := node.(*mark.CodeNode)
+			if !ok {
+				return
+			}
+			p.Text = strings.NewReplacer("&lt;", "<", "&gt;", ">", "&quot;", `"`, "&amp;", "&").Replace(p.Text)
+			if skipEntries > 0 {
+				skipEntries--
+				return
+			}
+
+			testFilePath := filepath.Join(tmpDir, fmt.Sprintf("%d.go", entryN))
+			testBinPath := filepath.Join(tmpDir, fmt.Sprintf("%d", entryN))
+			w, werr := os.Create(testFilePath)
+			if werr != nil {
+				panic(werr)
+			}
+			t.Execute(w, p)
+			w.Sync()
+			w.Close()
+			entryN++
+
+			cmdArgs := []string{"-s", "-w", "-l", testFilePath}
+			exec.Command("gofmt", cmdArgs...).Run()
+
+			cmdArgs = []string{"-w", testFilePath}
+			exec.Command("goimports", cmdArgs...).Run()
+
+			// Go build current directory
+			cmdArgs = []string{"build", "-o", testBinPath, testFilePath}
+			if cerr := exec.Command("go", cmdArgs...).Run(); cerr != nil {
+				fmt.Printf("Failed to validate code at %s:%d\n%s", docPath, p.Pos, p.Text)
+				os.Exit(255)
+			}
+			fmt.Printf("Successfully validated %s code at %s:%d\n", p.Lang, docPath, p.Pos)
```

Comment:
p.Pos is a little confusing, not clear what position it is referring to

```sh
Successfully validated go code at docs/API.md:7695
Successfully validated go code at docs/API.md:8261
Successfully validated go code at docs/API.md:8847
Successfully validated go code at docs/API.md:9279
Successfully validated go code at docs/API.md:10467
Successfully validated go code at docs/API.md:11698
Successfully validated go code at docs/API.md:13270
Successfully validated go code at docs/API.md:14553
Successfully validated go code at docs/API.md:15424
Successfully validated go code at docs/API.md:16389
Successfully validated go code at docs/API.md:17463
```

Reply:
A position is the position in file with number of characters. It doesn't equate to line number.

Reply:
in that case, I dont feel it adds any value, can we remove it?